src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala

```diff
@@ -178,16 +180,60 @@ object ZipAndJarSourcePathFactory extends ZipAndJarFileLookupFactory {
 
 final class FileBasedCache[T] {
   import java.nio.file.Path
-  private case class Stamp(lastModified: FileTime, fileKey: Object)
+  private case class Stamp(size: Long, signatures: ArrayBuffer[Long], fileKey: Object)
   private val cache = collection.mutable.Map.empty[Seq[Path], (Seq[Stamp], T)]
 
+  /**
+    * Read the CRCs of every zip entry if all of them succeed to be read.
+    *
+    * This method uses `java.util.jar` instead of `java.util.zip` utilities
+    * because the method `getCrc` in [[java.util.zip.ZipEntry]] returns 0 or
+    * -1 values.
+    *
+    * @param jarFile A jar file (which follows the zip specification format).
+    * @return A list of checksums for every zip entry.
+    */
+  def readJarCrcs(jarFile: Path): Option[ArrayBuffer[Long]] = {
+    try {
+      val jf = new java.util.jar.JarFile(jarFile.toFile)
+      try {
+        val entries = jf.entries()
+        val buffer = new ArrayBuffer[Long](32)
+        while (entries.hasMoreElements) buffer.+=(entries.nextElement().getCrc)
+        // It is very likely that a jar with no CRCs entries is ill-generated, assume so
+        if (buffer.isEmpty || buffer.forall(_ == -1L)) None else Some(buffer)
+      } finally jf.close()
+    } catch {
+      case _: IOException => None
+    }
+  }
+
+  /**
+    * Get a cached value or create it afresh if the value is missing or
+    * any of the given paths have been "modified".
+    *
+    * This implementation considers that a file has been modified if:
+    *   1. The file has changed its total size.
+    *   2. The CRCs present in the jar file are not the same (in the same order).
+    *   3. The file key (uniquely identifying a file) is different.
+    *
+    * Note that if there's a IO error while reading the CRCs in the cached paths,
+    * this implementation fallbacks to the use of the last modified time. This
+    * also happens when jars have no CRCs -- it is extremely unlikely that a
+    * classpath entry has no CRC, so if that happens we assume the jar is
+    * either corrupt or ill-generated, and we use a safer change detection strategy.
```

Comment:
not sure the fallback is safer. Maybe simpler though

I am not convinced that if the jar has no crcs it should even be considered equal. Surely that would be safer

Reply:
> I am not convinced that if the jar has no crcs it should even be considered equal. Surely that would be safer
  
It's not considered equal, it uses the last modified time instead, which is the current strategy.

Reply:
If the last modification date was sufficient then all of the crc scanning, and this whole PR is not needed, as the implementation that it overrides is 'safer' then its replacement.
As I understand it you want to scan the CRC because the last modification date is not sufficient, so therefore it cannot be safer to just use an inferior check

my other point is that IMO a "corrupt or ill-generated" jar should not even be equal to itself - the same strategy used for directories, so we should not cache it. I think that later we will have to have a policy that defines this behavior. Other may/will have a different opinion which depend on their build infrastructure, which is why it will have to be policy based

Reply:
I think we can provide an implementation that reconciles all the reasonable use cases in many build tools and environments.
  
Re modication time, I've addressed that concern in another comment. Re corrupt or ill-generated jar, I'm skeptical we should not cache them. Corrupt or ill-generated jars is a bad notion I've come up with to describe jars whose CRCs are zero. These jars may very well be valid from a tool's or user's point of view, and can be generated by so many clients that we cannot just "not cache it", as I see it.